File: src/lib/stores/friendlyChat.ts
import { create } from 'zustand'
import type { FriendlyChatSettings, FriendlyChatMessage } from '@/lib/types/friendlyChat'
import {
  getFriendlyChatSettings,
  updateFriendlyChatSettings,
  getFriendlyChatHistory,
  triggerFriendlyChat
} from '@/lib/client/apiClient'

interface FriendlyChatState {
  settings: FriendlyChatSettings
  messages: FriendlyChatMessage[]
  loading: boolean
  error: string | null

  // Actions
  fetchSettings: () => Promise<void>
  updateSettings: (settings: Partial<FriendlyChatSettings>) => Promise<void>
  fetchHistory: (limit?: number, offset?: number) => Promise<void>
  triggerChat: () => Promise<string | null>
  addMessage: (message: FriendlyChatMessage) => void
}

const defaultSettings: FriendlyChatSettings = {
  enabled: false,
  interval: 20,
  dataWindow: 20,
  enableSystemNotification: true,
  enableLive2dDisplay: true
}

export const useFriendlyChatStore = create<FriendlyChatState>((set, get) => ({
  settings: defaultSettings,
  messages: [],
  loading: false,
  error: null,

  fetchSettings: async () => {
    set({ loading: true, error: null })
    try {
      console.log('[FriendlyChat] Fetching settings...')
      const response: any = await getFriendlyChatSettings()

      if (response.success && response.data) {
        const settings: FriendlyChatSettings = {
          enabled: response.data.enabled ?? defaultSettings.enabled,
          interval: response.data.interval ?? defaultSettings.interval,
          dataWindow: response.data.dataWindow ?? defaultSettings.dataWindow,
          enableSystemNotification: response.data.enableSystemNotification ?? defaultSettings.enableSystemNotification,
          enableLive2dDisplay: response.data.enableLive2dDisplay ?? defaultSettings.enableLive2dDisplay
        }
        set({ settings, loading: false })
      } else {
        throw new Error('Failed to fetch settings')
      }
    } catch (error) {
      console.error('Failed to fetch friendly chat settings:', error)
      set({ error: (error as Error).message, loading: false, settings: defaultSettings })
    }
  },

  updateSettings: async (updates: Partial<FriendlyChatSettings>) => {
    set({ loading: true, error: null })
    try {
      console.log('[FriendlyChat] Updating settings:', updates)
      const response: any = await updateFriendlyChatSettings(updates)

      if (response.success && response.data) {
        const settings: FriendlyChatSettings = {
          enabled: response.data.enabled ?? defaultSettings.enabled,
          interval: response.data.interval ?? defaultSettings.interval,
          dataWindow: response.data.dataWindow ?? defaultSettings.dataWindow,
          enableSystemNotification: response.data.enableSystemNotification ?? defaultSettings.enableSystemNotification,
          enableLive2dDisplay: response.data.enableLive2dDisplay ?? defaultSettings.enableLive2dDisplay
        }
        set({ settings, loading: false })
      } else {
        throw new Error(response.message || 'Failed to update settings')
      }
    } catch (error) {
      console.error('Failed to update friendly chat settings:', error)
      set({ error: (error as Error).message, loading: false })
      // Re-fetch to ensure consistency
      await get().fetchSettings()
    }
  },

  fetchHistory: async (limit = 20, offset = 0) => {
    set({ loading: true, error: null })
    try {
      console.log('[FriendlyChat] Fetching history...', { limit, offset })
      const response: any = await getFriendlyChatHistory({ limit, offset })

      if (response.success && response.data) {
        set({ messages: response.data.messages || [], loading: false })
      } else {
        throw new Error('Failed to fetch history')
      }
    } catch (error) {
      console.error('Failed to fetch friendly chat history:', error)
      set({ error: (error as Error).message, loading: false, messages: [] })
    }
  },

  triggerChat: async () => {
    set({ loading: true, error: null })
    try {
      console.log('[FriendlyChat] Triggering immediate chat...')
      const response: any = await triggerFriendlyChat()

      set({ loading: false })

      if (response.success && response.data?.chatMessage) {
        return response.data.chatMessage
      } else {
        console.warn('Failed to generate chat message:', response.message)
        return null
      }
    } catch (error) {
      console.error('Failed to trigger friendly chat:', error)
      set({ error: (error as Error).message, loading: false })
      return null
    }
  },

  addMessage: (message: FriendlyChatMessage) => {
    set((state) => ({
      messages: [message, ...state.messages]
    }))
  }
}))

File: backend/handlers/friendly_chat.py
"""Friendly Chat handlers."""

from __future__ import annotations

from typing import Dict, Any

from core.settings import get_settings
from services.friendly_chat_service import get_friendly_chat_service

from . import api_handler
from models.requests import (
    UpdateFriendlyChatSettingsRequest,
    GetFriendlyChatHistoryRequest,
)


@api_handler(method="GET")
async def get_friendly_chat_settings() -> Dict[str, Any]:
    """Get friendly chat configuration.

    Returns the current settings for the friendly chat feature including
    interval, data window, and notification preferences.
    """
    settings = get_settings()
    chat_settings = settings.get_friendly_chat_settings()

    return {
        "success": True,
        "data": chat_settings,
    }


@api_handler(body=UpdateFriendlyChatSettingsRequest)
async def update_friendly_chat_settings(
    body: UpdateFriendlyChatSettingsRequest,
) -> Dict[str, Any]:
    """Update friendly chat configuration.

    Updates the friendly chat settings and restarts the service if needed.
    """
    settings = get_settings()
    chat_service = get_friendly_chat_service()

    # Update settings
    updated = settings.update_friendly_chat_settings(body.model_dump(exclude_none=True))

    # Restart service based on enabled status
    if updated.get("enabled", False):
        await chat_service.stop()  # Stop if running
        await chat_service.start()  # Start with new settings
    else:
        await chat_service.stop()

    return {
        "success": True,
        "message": "Friendly chat settings updated",
        "data": updated,
    }


@api_handler(body=GetFriendlyChatHistoryRequest)
async def get_friendly_chat_history(
    body: GetFriendlyChatHistoryRequest,
) -> Dict[str, Any]:
    """Get friendly chat message history.

    Returns a paginated list of previously generated chat messages.
    """
    chat_service = get_friendly_chat_service()
    history = await chat_service.get_chat_history(
        limit=body.limit,
        offset=body.offset,
    )

    return {
        "success": True,
        "data": {
            "messages": history,
            "count": len(history),
        },
    }


@api_handler(method="POST")
async def trigger_friendly_chat() -> Dict[str, Any]:
    """Manually trigger a friendly chat message generation.

    Generates and sends a chat message immediately based on recent activities.
    """
    chat_service = get_friendly_chat_service()
    message = await chat_service.trigger_immediate_chat()

    if message:
        return {
            "success": True,
            "message": "Chat message generated",
            "data": {"chat_message": message},
        }
    else:
        return {
            "success": False,
            "message": "Failed to generate chat message (no recent activities or LLM error)",
        }
